Mono.Debugging/Mono.Debugging.Client/SourceLink.cs

```diff
@@ -1,12 +1,19 @@
-﻿using System;
+using System;
+using System.Collections.Generic;
 using System.IO;
+using System.Linq;
+using System.Text;
 
 namespace Mono.Debugging.Client
 {
 	[Serializable]
 	public class SourceLink
 	{
+		// Keep the / character to use as a path separator to help group related symbols
+		static readonly HashSet<char> invalids = new HashSet<char>(Path.GetInvalidFileNameChars ().Except (new char[] { '/' }));
```

Comment:
I think an array here is probably going to be faster than a hashset.

Reply:
But it's not a big deal.

Reply:
it's a lookup table that is used repeatedly, so hashset is a better choice here

Reply:
Have you _profiled_ it? I assume not, so here are some results that [say the opposite](https://github.com/Therzok/BenchPlayground/blob/8377c906a5a4aa2f1f7472b67a0e15c3dfcf713e/BenchPlayground/Benchmarks/System.Collections/ArrayVsHashSetStrings.md).

You have to compute a hashcode to get a bucket then do equality comparison for a HashSet. For simple arrays like `char[]`, `Array.IndexOf` does a SIMD scan.

Reply:
I actually assumed that System.IO.Path.GetInvalidFileNameChars () would contain many more than 2 items :)

Reply:
but also, your profile results are only valid for the exact number of items in the array that you use (5)

Reply:
> I actually assumed that System.IO.Path.GetInvalidFileNameChars () would contain many more than 2 items :)
> 
> ```
> csharp> System.IO.Path.GetInvalidFileNameChars ()    
> { '\x0', '/' } 
> ```
> 
> If I'd realised that there were only 2 items, I would have never used a HashSet. No need for profile results here :D

This actually makes me question whether or not MakeValidFileName(uri) could just be => uri.Host + uri.PathAndQuery; as I don't think \x0 is a possible character in uri.Host or uri.PathAndQuery

Reply:
debugger-libs should still run on Windows, which returns a bit different values:
https://sharplab.io/#v2:C4LgTgrgdgPgAgJgAwFgBQiCM65IARyYB0AkgPIDc66AZgPZgCmAhgMYAWeAFAG7Nh4OeAJZQ8ABWbB2RAOKNgJKHwA2wgCYAxYSsYA5ZgFtGAYXb8AzlwCU1vAG90eZ4QCcXAET2kAXweYfDwAaQXYQri4AIwBPYEZrDlsKPHQfdCA=

Reply:
> but also, your profile results are only valid for the exact number of items in the array that you use (5)

It's valid for a lot of values until 64, from what I recall, there are other benchmarks in that repo that are relevant.

Reply:
Indeed, MakeValidFileName is not even needed at all

Reply:
> debugger-libs should still run on Windows, which returns a bit different values:
> https://sharplab.io/#v2:C4LgTgrgdgPgAgJgAwFgBQiCM65IARyYB0AkgPIDc66AZgPZgCmAhgMYAWeAFAG7Nh4OeAJZQ8ABWbB2RAOKNgJKHwA2wgCYAxYSsYA5ZgFtGAYXb8AzlwCU1vAG90eZ4QCcXAET2kAXweYfDwAaQXYQri4AIwBPYEZrDlsKPHQfdCA=

grr.. I forgot it needs to work on Windows. Looks like 41 results. I would be very surprised if Array came out on top here

Reply:
Even on Windows, MakeValidFileName is probably not needed. I don't think there are any URI characters that are invalid file characters.